src/main/java/com/cloudbees/plugins/credentials/CredentialsDescriptor.java

```diff
@@ -192,4 +257,339 @@ public String getIconClassName() {
+                o = ((CredentialsStoreAction.CredentialsWrapper) o).getStore().getContext();
+            } else if (o instanceof CredentialsStoreAction.DomainWrapper) {
+                o = ((CredentialsStoreAction.DomainWrapper) o).getStore().getContext();
+            } else if (o instanceof Descriptor && i == 1) { // URL is /descriptorByName/...
+                // we need to try an infer from the Referer as this is likely a doCheck or a doFill method
+                String referer = request.getReferer();
+                String rootPath = request.getRootPath();
+                if (referer != null && rootPath != null && referer.startsWith(rootPath)) {
+                    // strip out any query portion of the referer URL.
+                    String path = URI.create(referer.substring(rootPath.length())).getPath().substring(1);
+                    // TODO have Stapler expose a method that can walk a path and produce the ancestors and use that
+
+                    // what now follows is an example of a realy evil hack, consequently this means...
+                    //
+                    //                         7..       ,
+                    //                      MMM.          MMM.
+                    //                     MMMMM        .MMMMMM
+                    //                    MMMM.           MMMMM.
+                    //                  OMMM                 MMZ
+                    //                  MMM                    MM
+                    //                .MMMM    $.     .       .MM,
+                    //                MMMMM    MMM   MM        MMM
+                    //               .MMMMM. MMMMD  8MMM.     MMMM
+                    //               MMMMMMM.MMMM    MMMMM.  MMMMMM
+                    //               MMMMMMMM.M .     MMM.  MMMMMMM
+                    //               MMMMMMMMM.            MMMMMMMM
+                    //               MMMMMMMMM . ..     MMMMMMMMMMM
+                    //               MMMMMMMM  IMMMM  Z.MMMMMMMMMM ,
+                    //               .MMMMMMM   .M:M   MMMMMMMMMMM M
+                    //              I MMMMMMM.         MMMMMMMMMO  M
+                    //           MMMM  MMMMMMM       .MMMMMMMMM.   .
+                    //         :MMMMMM.MMMMMMM.      MMMMMMMM   .MMMM
+                    //         MMMMMMMMMMMMMMMM      MMMMMMM   MMMMMMMM
+                    //        MMMMMMMMM.MMMMMMM      MMMMMMM MMMMMMMMMM.
+                    //        MMMMMMMMMMMMMMMM?      MMMMMM MMMMMMMMMMM
+                    //        MMMMMMMMMM  .  .       MMMMMIMMMMMMMMMMMM.
+                    //        MMMMMMMMMM               .. :MMMMMMMMMMMM.
+                    //        DMMMMMMMMMM                 MMMMMMMMMMMMM.
+                    //         MMMMMMMMMM.M.              MMMMMMMMMMMM.
+                    //           MMMMMM,                    ....
+                    //
+                    //                   I AM A SAD PANDA
```

Comment:
🐜 You should feed your panda better :) But it would be great to have an issue link